api/database/boltdb/db.go

```diff
@@ -65,17 +65,63 @@ func (connection *DbConnection) IsEncryptedStore() bool {
 // NeedsEncryptionMigration returns true if database encryption is enabled and
 // we have an un-encrypted DB that requires migration to an encrypted DB
 func (connection *DbConnection) NeedsEncryptionMigration() bool {
+
+	// Cases:  Note, we need to check both portainer.db and portainer.edb
+	// to determine if it's a new store.   We only need to differentiate between cases 2,3 and 5
+
+	// 1) portainer.edb + key     => False
+	// 2) portainer.edb + no key  => ERROR Fatal!
+	// 3) portainer.db  + key     => True  (needs migration)
+	// 4) portainer.db  + no key  => False
+	// 5) NoDB (new)    + key     => False
+	// 6) NoDB (new)    + no key  => False
+	// 7) portainer.db & portainer.edb => ERROR Fatal!
+
+	// If we have a loaded encryption key, always set encrypted
 	if connection.EncryptionKey != nil {
-		dbFile := path.Join(connection.Path, DatabaseFileName)
-		if _, err := os.Stat(dbFile); err == nil {
-			return true
+		connection.SetEncrypted(true)
+	}
```

Comment:
I had another look and track those calls and I don't think they actually do a useful job.
The decision to run migration of not is based on the `NeedsEncryptionMigration()`, which uses `EncryptionKey` directly. 
`IsEncryptedStore` is used to determine db file name and nothing stops us from looking at the `EncryptionKey` there.
`SetEncrypted` is used to set an internal field that is used during key fetching, which in its turn could be replaces with call to `EncryptionKey` as well.
That may reduce the interface of the connection, which is a good thing, i think.

Reply:
I tried a bunch of different things without a correctly working solution.   Personally, I think the migration should be done in the db level itself, but that requires a lot more effort.  I think we should just go with how it works currently.   It can be refactored at a later date.